design/unnumbered-bgp.md

```diff
@@ -0,0 +1,498 @@
+
+Total number of neighbors 1
+
+IPv6 Unicast Summary (VRF default):
+BGP router identifier 172.19.0.3, local AS number 65000 vrf-id 0
+BGP table version 1
+RIB entries 1, using 96 bytes of memory
+Peers 1, using 13 KiB of memory
+Peer groups 1, using 64 bytes of memory
+
+Neighbor                  V         AS   MsgRcvd   MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd   PfxSnt Desc
+unnumbered-p2p-peer(net0) 4      65004       512       512        1    0    0 00:25:15            0        1 TOR
+
+Total number of neighbors 1
+```
+### Topics
+
+#### Loopback IPv4 address
+
+Linux sends and receives IPv4 packets over an interface that does not have an IPv4 address.
+It does not look we need to add address in the loopback.
+
+#### RA enabled on both ends of the link
+
+According to the design of unnumber BGP peering, we need to enable RAs.
+
+#### BGP RouterID
+
+//TODO is there a requirement for that?
+
+### Traffic forwarding
+
+FRR advertises IPv4 prefixes with IPv6 next hops using encoding specified in RFC 5549 (now RFC 8950).
+How it works is outside the scope of the doc, but the note to keep is that we can test unnumber BGP peering
+in IPv4 setup (as long as the IPv6 stack/LLAs are not disabled).
+
+```
+> docker exec -it unnumbered-p2p-peer vtysh -c 'show bgp ipv4'
+BGP table version is 3, local router ID is 11.11.11.254, vrf id 0
+Default local pref 100, local AS 65004
+Status codes:  s suppressed, d damped, h history, * valid, > best, = multipath,
+               i internal, r RIB-failure, S Stale, R Removed
+Nexthop codes: @NNN nexthop's vrf id, < announce-nh-self
+Origin codes:  i - IGP, e - EGP, ? - incomplete
+RPKI validation codes: V valid, I invalid, N Not found
+
+    Network          Next Hop            Metric LocPrf Weight Path
+ *= 5.5.5.5/32       eth02                    0             0 65000 i
+ *=                  eth01                    0             0 65000 i
+ *>                  eth00                    0             0 65000 i
+ *> 200.100.100.0/24 0.0.0.0(unnumbered-p2p-peer)
+                                             0         32768 ?
+
+Displayed  2 routes and 4 total paths
+> docker exec -it unnumbered-p2p-peer ip route
+5.5.5.5 nhid 21 proto bgp metric 20
+        nexthop via inet6 fe80::dcad:beff:feff:1160 dev eth00 weight 1
+        nexthop via inet6 fe80::dcad:beff:feff:1161 dev eth01 weight 1
+        nexthop via inet6 fe80::dcad:beff:feff:1162 dev eth02 weight 1
+```
+
+### The API Extension
+
+We have three option:
+
+#### Option 1. Re-Use Address field
+
+We can re-use `Address string` to either be IP or interface, if not valid IP we
+indicate Unnumbered BGP peering. This looks the minimal API, but it will not
+have clean error path. E.g. if the user input is not valid IP, the
+implementation will shadow all the existing errors and will fail in the
+runtime.
+
+#### Option 2. Add a new Interface field
```

Comment:
I think this is the most valid one, and one that matches FRR the most in a sense (since it seems there's a distinction between interfaces and addresses) without locking us. Also, the docs don't need to be as complicated (stating all of the cases) as it's simply following the what frr is doing, that is we can just shortly say, if address is not specified the peering happens over the interface (unnumbered), if address is specified it is assumed the it is lla.

Reply:
+1 on this

Reply:
ACK then, will put that as but we need to decide if to further https://github.com/metallb/metallb/issues/2423
and have this 
````
// 1. if address and no interface then as normal
// 2. if non valid address and no interface then fail
// 3. if address is LLA and interface not empty, then LLA support
// == extra FRR directive `neighbor fe80::a876:4dff:fe77:408 interface eth0`
// 4. If address is LLA and empty interface, then fail
// 5. If address is empty and interface has value then the BGP unnumber peering config will take place
// ==  modify FRR directive `neighbor PEER interface remote-as <ASN>`
// && replace any address directive to use interface e.g.  `neighbor net0 timers 180 540`
// && enable RAs on this interface
// 6. If address has value (non-LLA) and interface non empty, then fail
``` 
as the end API

Reply:
not sure I understand - implementation wise the issue will be solved for free(?) and documentation wise it seems sufficient to document the `Interface` field something like "if specified without address then peering happens over the interface (unnumbered), if together with address then it is assumed to be LLA"?